### PR TITLE
clientv3: Fix maintenance APIs to directly dial grpc endpoints directly

### PR DESCRIPTION
**Pull request by jpbetz**
This fixes `clientv3.Dial` function to directly dial a single endpoint. 

Conveniently, it seems this can be fixed by removing special casing logic from clientv3. Previously clientv3 had code in `dialSetupOpts` that would attempt to circumvent the endpoint the balancer had selected when performing a `net.Dial`, but we had accidentally disabled this code path in the grpc new load balancer upgrade (with a condition that always returned false), and attempts to fix it to properly handle all use cases error prone, attempts to fix it could easily be applied it too broadly such that the balancer became unintentionally disabled when it should not. Instead of using that approach, this PR simply dials with a non-load balanced grpc target for all connections established via `clientv3.Dial`.

Fixes https://github.com/coreos/etcd/issues/9919

Example etcdctl status call after applying this to master:

```
bin/etcdctl endpoint status --cluster
http://127.0.0.1:2379, 8211f1d0f64f3269, 3.3.0+git, 422 MB, false, 14, 118988, 118988,                   
http://127.0.0.1:22379, 91bc3c398fb3c146, 3.3.0+git, 423 MB, false, 14, 118988, 118988,              
http://127.0.0.1:32379, fd422379fda50e48, 3.3.0+git, 423 MB, true, 14, 118988, 118988,
```

cc @gyuho @jingyih

### CODE REVIEW

**Comment by gyuho** [comment]
@jpbetz This fix looks good. Let's merge when CIs green. Thanks!

**Comment by jpbetz** [comment]
e2e show a problem with client tls.  I’ll sort that out shortly

**Comment by jpbetz** [comment]
Added a couple minor convenience utilities to the endpoint resolver for direct endpoints and all tests are passing locally now. If CI passes this is ready for final review.

**Comment by gyuho** [comment]
Can you also rebase from current master? Thanks!

**Comment by jpbetz** [comment]
@gyuho Rebased.

I'm not 100% happy with how I've had to add a `DirectResolver` to `clientv3/balancer/resolver/endpoint.go`.  It should have been possible to use the builtin grpc passthrough resolver and even the direct resolver shouldn't need to be written the way it is where the scheme is passed in as part of the 'authority' part of the grpc target, .e.g `endpoint://direct:https/localhost:2379`. But it's the only format that works for all our test cases.  If I try to use a target format like `endpoint://direct/https://localhost:2379` the TLS e2e test fails, which suggests that something about our TLS cert handling is inspecting the grpc target and distinguishes between the 'https://localhost:2379' endpoint part and the 'localhost:2379' endpoint part, but I haven't found out where.

**Comment by gyuho** [comment]
> TLS cert handling is inspecting the grpc target

Wish there be cleaner ways. I will research more around this change as well.

**Comment by jpbetz** [comment]
@gyuho Sounds good, I'll continue to dig in as well. Let's not merge this until we're got something we're both happy with.

**Comment by jpbetz** [comment]
Switched to using the builtin grpc 'passthrough' resolver when dialing a single endpoint directly.

The main complication with this approach is that for x509 cert validation, grpc uses this logic in `DialContext` to infer the authority to check:

```
creds := cc.dopts.copts.TransportCredentials
     if creds != nil && creds.Info().ServerName != "" {
         cc.authority = creds.Info().ServerName
     } else if cc.dopts.insecure && cc.dopts.copts.Authority != "" {
         cc.authority = cc.dopts.copts.Authority
     } else {
         // Use endpoint from "scheme://authority/endpoint" as the default
         // authority for ClientConn.
         cc.authority = cc.parsedTarget.Endpoint
     }
 ...
 }
```
Which results in a client connection's `authority` being set to different values depending on a few different conditions.  It works for our balancer because we happen to have the endpoint part of the target set to a host value, which is a valid authority for TLS purposes.  This is also why for direct host connections, endpoints like `endpoint://direct/https://localhost:2379` work (since the endpoint is just the hostname), but `endpoint://direct/https://localhost:2379` does not (the authority becomes `https://localhost:2379` which does work as needed by TLS).

I've worked around by adding a `dialDirectCert` function that configures a cert explicitly using the `OverrideServerName` function provided by grpc's `TransportCredentials` interface.  I've also restructured how we dial slightly to make cert handling more clear/explicit.

File: clientv3/client.go [hunk 1]
```diff
@@ -291,8 +275,9 @@ func (c *Client) dialSetupOpts(target string, dopts ...grpc.DialOption) (opts []
 }
 
 // Dial connects to a single endpoint using the client's config.
-func (c *Client) Dial(endpoint string) (*grpc.ClientConn, error) {
-	return c.dial(endpoint)
+func (c *Client) Dial(ep string) (*grpc.ClientConn, error) {
+	creds := c.directDialCreds(ep)
+	return c.dial(fmt.Sprintf("passthrough:///%s", ep), creds)
```
Comment (gyuho) [inline clientv3/client.go]: Can we add a short comment on `passthrough:///`?

**Comment by gyuho** [comment]
@jpbetz Never thought of using `passthrough`, and this seems much cleaner than I thought it would be. But I think functional testing is failing from this. I will investigate this change today.

> spell: hostname -> host name?

In the meantime, can we fix this test failure? We can add `hostname` to `.words` file.

**Comment by jpbetz** [comment]
Thanks for having a quick look @gyuho.  I'll clean this up and dig in to the functional tests.

File: clientv3/client.go (hunk 1 above)
Comment (jpbetz) [reply clientv3/client.go]: Sounds good, I'll add one now.

Comment (jpbetz) [reply clientv3/client.go]: Certainly, I've clarified why we use the passthrough resolver here.

**Comment by jpbetz** [comment]
Simplified the code changes and fixed a subtle bug in the cred configuration logic. Functional tests are now passing locally. I'll keep an eye on CI.

File: clientv3/integration/dial_test.go [hunk 2]
```diff
@@ -79,20 +79,15 @@ func TestDialTLSNoConfig(t *testing.T) {
...
-	}
+	defer func() {
+		if c != nil {
+			c.Close()
+		}
+	}()
 	if !isClientTimeout(err) {
 		t.Fatalf("expected dial timeout error, got %v", err)
 	}
+
```
Comment (gyuho) [inline clientv3/integration/dial_test.go]: Remove this line?

**Pull request merged**